--- patches/memcache/dmemcache.inc	2016-05-09 15:30:35.000000000 +0200
+++ contrib/memcache/dmemcache.inc	2016-05-09 15:30:43.000000000 +0200
@@ -394,7 +394,12 @@
 
   if ($collect_stats) {
     foreach ($multi_stats as $key => $value) {
-      $multi_stats[$key] = isset($results[$key]) ? TRUE : FALSE;
+      // If we are using the broken version of memcache library, we need to
+      // prepend the key prefix to get correct statistics for get/set_multiple.
+      if (variable_get('memcache_broken_lib', FALSE)) {
+        $result_key = variable_get('memcache_key_prefix', '') . $key;
+      }
+      $multi_stats[$key] = isset($results[$result_key]) ? TRUE : FALSE;
     }
   }
 
@@ -406,6 +411,15 @@
     if (is_object($value) && !empty($value->multi_part_data)) {
       $value = _dmemcache_get_pieces($value->data, $value->cid, $bin, $mc);
     }
+    // If we are using the broken version of memcache library, we need to remove
+    // the prefix from the stored key, because it's there twice, by mistake.
+    // Otherwise get/set_multiple operation doesn't work correctly.
+    if (variable_get('memcache_broken_lib', FALSE)) {
+      $prefix = variable_get('memcache_key_prefix', '');
+      if (substr($key, 0, strlen($prefix)) == $prefix) {
+        $key = substr($key, strlen($prefix));
+      }
+    }
     $cid_results[$cid_lookup[$key]] = $value;
   }
 
